src/Types.elm

```diff
@@ -2,18 +2,73 @@ module Types exposing (..)
 
 type Tree a = Node a (List (Tree a))
 
-type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole | TokVar String
-             | TokConstInt Int | TokConstBool Bool | TokInvalid | TokBackSlash | TokArrow | TokEnd
+type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole |
+              TokVar String | TokHasType | TokConstInt Int | TokConstBool Bool |
+              TokInvalid | TokBackSlash | TokArrow | TokEnd | TokTypeName String
 
 type TokTSC = TTSCInt TokTSCInt | TTSCBool TokTSCBool
 type TokTSCInt = TokPlus | TokMinus | TokTimes
 type TokTSCBool = TokEq | TokAnd | TokOr
 
 type Const = CBool Bool | CInt Int
 type Term = CTerm Const | VTerm String | Plus Term Term | Minus Term Term | Times Term Term
-            | Eq Term Term | And Term Term | Or Term Term | Lam Term Term | App Term Term
+            | Eq Term Term | And Term Term | Or Term Term | Lam String Term | App Term Term
             | MissingInt | MissingBool | Missing | EmptyTree
 
+-- V(alue)Type is a type that a TreeAssembly term can evaluate to
+{-
+  TAny may be used when the type can be either int or bool (discuss)
+  TNone is used as a default value for types we haven't determined yet (similar to EmptyTree)
+-}
+type VType = TBool | TInt | TFun VType VType | TAny | TNone
+
 type alias Var =
   { name: String
-  , term: Term }
+  , term: Term
+  , vtype: VType}
+
+{-
+  Returns the type signature of basic language constructs (if it's not relevant, return TNone)
+-}
```

Comment:
I think this is a case where we could use Maybe to avoid a `TNone`: i.e. `getTypeSignature` could be `Term -> Maybe VType`

Reply:
Fixed!